tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -866,19 +871,20 @@ async def hover(ls: ChapelLanguageServer, params: HoverParams):
         text_doc = ls.workspace.get_text_document(params.text_document.uri)
 
         fi, _ = ls.get_file_info(text_doc.uri)
-        segment = fi.get_use_segment_at_position(params.position)
+
+        segment = fi.get_use_or_def_segment_at_position(params.position)
         if not segment:
             return None
-        resolved_to = segment.resolved_to
-        node_fi, _ = ls.get_file_info(resolved_to.get_uri())
 
-        signature = get_symbol_signature(resolved_to.node)
-        docstring = chapel.get_docstring(resolved_to.node, node_fi.siblings)
+        node_fi, _ = ls.get_file_info(segment.get_uri())
+
+        signature = get_symbol_signature(segment.node, eval_expressions=ls.eval_expressions, resolve=ls.use_resolver)
```

Comment:
Since this now depends on the state of `ls`, consider defining a `ls.get_symbol_signature` method that calls out to the (plain) `get_symbol_signature`.

More completely, consider wrapping the functions from `symbol_signature` into a class: `SymbolSignature`. The stuff that's _always_ displayed can be computed in the constructor, and additional stuff can be computed via `compute_type` and `compute_value` methods. Finally, a `__str__` (or regular method) can be used to extract the ... markdown? ... from the SymbolSignature. Then,

```Python
def get_symbol_signature(self: ChapelLanguageServer, node: chapel.AstNode):
    sig = get_symbol_signature(node)
 
    if self.use_resolver:
        sig.compute_type()
        if self.eval_expressions:
            sig.compute_value()
    return sig
```

This way, you avoid polluting the `get_symbol_signature` methods and its callees with unnecessary boolean arguments that represent the configuration state, and you keep that code re-usable in case we ever want to move it to the main module.

Reply:
So because of the recursive nature of `get_symbol_signature`, this isn't a simple change. As a first step, I wrapped `get_symbol_signature` in SymbolSignature, which has the same API as you describe. However its just a nice wrapper over the function with passing arguments around, and can cause necessary recomputation. So still a work in progress

Reply:
That's unfortunate. That signature code needs some attention as is, IMO.